### ISSUE TITLE
pi07 video encoder is not faithful to the MEM paper (architecture + residual)

### ISSUE THREAD

**Issue by shuheng-liu** [TensorAuto/OpenTau#270]
`src/opentau/policies/pi07/low_level/video_encoder.py` is documented as implementing the video encoder from Torne, Pertsch, Walke et al. *"MEM: Multi-Scale Embodied Memory for VLA Models"* (Section III-C, Appendix C). It is **not faithful to the paper** in two places. Both should be fixed.

## Bug 1: Architecture diverges from Appendix C; `T == 1` skip is a workaround, not a fix

Lines 319–325:

```python
if t == 1:
    return self._spatial_block_forward(hidden_states, attention_mask, output_attentions)
```

The paper claims that the boundary condition `e(0) = 0` alone is sufficient for single-frame invariance. With our current **stacked-residual** layout (a temporal sub-block with its own residual, followed by a standard spatial sub-block), this is mathematically false. For `T = 1` and `is_causal=True`, the temporal attention output is

```
out_proj(softmax(qk^T) v) = out_proj(v_proj(LN1(x))) ≠ 0
```

so the residual `x + out_proj(v_proj(LN1(x)))` is not equal to `x`. We currently paper over this by skipping the temporal block entirely at `T = 1`, but that is not what the paper describes — Appendix C does not contain any such conditional.

The faithful reading of Appendix C, Eq. 3,

```
α(S = {1, ..., N}, T = {}) [ α(S = {}, T = {1, ..., T}) [ẑ] ]
```

is a **fused factorized attention** — a single attention operation whose weights factor as `α_total[(p,t),(p',t')] = α_spatial(p, p') · α_temporal(t, t')`, applied once with one residual. Under this formulation, `T = 1` makes `α_temporal(0, 0) = 1`, so `α_total = α_spatial` exactly, and the layer collapses to standard ViT spatial attention with no special-case needed.

**Fix:** restructure `SpaceTimeEncoderLayerWrapper.forward` as a single fused space-time attention block (one residual, factorized weights), and remove the `if t == 1` short-circuit. After this change, single-frame invariance holds via `e(0) = 0` alone — matching the paper's actual claim.

## Bug 2: Residual carries `z` instead of `ẑ = z + e(t)`

Lines 343–350:

```python
x_pe = x + pe                                              # ẑ = z + e(t)
t_norm = self.layer_norm1(t_in)                            # LN over x_pe
t_out = self._temporal_attn(t_norm, ...)
t_res = rearrange(x, "b t n d -> (b n) t d") + t_out       # residual on x, NOT x_pe
```

Appendix C defines `ẑ = z + e(t)` and then says "we follow the standard computation of a transformer layer." A standard transformer block carries the LN-input through the residual, so the residual should be on `ẑ`, not `z`. Our current code uses `e(t)` only as a transient Q/K modulation that does not propagate through the residual stream. This is not what the paper describes.

**Fix:** change the residual to use `x_pe` (i.e., `ẑ`):

```python
t_res = rearrange(x_pe, "b t n d -> (b n) t d") + t_out
```

Once Bug 1 is fixed (single fused attention, single residual), the residual is straightforwardly on `ẑ`, and this issue resolves naturally.

## Why both fixes are coupled

Removing the `T == 1` skip without restructuring the attention will break single-frame invariance and prevent `pi05` checkpoints from loading as drop-in replacements. Restructuring to fused factorized attention is the change that lets us delete the workaround **and** get single-frame invariance for free, **and** carry `ẑ` through the residual without accumulating positional drift across spacetime layers (since there is now one residual per layer, not one per sub-block).

## Suggested implementation steps

- [ ] Replace the two stacked sub-blocks in `SpaceTimeEncoderLayerWrapper.forward` with a single fused space-time attention: compute `Q`, `K`, `V` from `LN(ẑ)`, build factorized attention weights `α_spatial · α_temporal` (with the existing causal time mask + padding mask), apply to `V`, then `out_proj`, then **one** residual on `ẑ`.
- [ ] Delete the `if t == 1: ...` short-circuit (lines 319–325).
- [ ] Update the module docstring's "Key properties" section to reflect that single-frame invariance now follows from `e(0) = 0` directly, with no special-case code.
- [ ] Verify `test_pi07_video_encoder_cpu.py::test_state_dict_keys_unchanged_after_wrapping` still passes (no new params should be introduced).
- [ ] Add a numerical test: forward a single-frame batch through the wrapped encoder and the unwrapped SigLIP, assert allclose at fp32.

## Files

- `src/opentau/policies/pi07/low_level/video_encoder.py` (lines 68–354)

**Comment by shuheng-liu** [comment]
Already fixed
